src/Microsoft.OpenApi.OData.Reader/Generator/OpenApiSchemaGenerator.cs

```diff
@@ -660,58 +642,64 @@ private static IOpenApiAny CreateStructuredTypePropertiesExample(ODataContext co
-                    IEdmTypeReference elementType = edmTypeReference.AsCollection().ElementType();
-                    array.Add(GetTypeNameForExample(context, elementType));
-                    return array;
+            if (edmTypeReference.IsBoolean())
+            {
+                return true;
+            }
+            else
+            {
+                if (schema is OpenApiSchemaReference { Reference.Id: not null } reference && !string.IsNullOrEmpty(reference.Reference.Id))
+                {
+                    return reference.Reference.Id;
+                }
+                else
+                {
+                    return schema.Type.ToIdentifier() ??
+                        (schema.AnyOf ?? Enumerable.Empty<IOpenApiSchema>())
+                        .Union(schema.AllOf ?? Enumerable.Empty<IOpenApiSchema>())
+                        .Union(schema.OneOf ?? Enumerable.Empty<IOpenApiSchema>())
+                        .FirstOrDefault(static x => !string.IsNullOrEmpty(x.Format))?.Format ?? schema.Format;
+                }
+            }
+        }
 
-                case EdmTypeKind.Untyped:
-                    return new OpenApiObject();
+        private static JsonNode GetTypeNameForExample(ODataContext context, IEdmTypeReference edmTypeReference, OpenApiDocument document)
+        {
+            return edmTypeReference.TypeKind() switch
+            {
+                // return new OpenApiBinary(new byte[] { 0x00 }); issue on binary writing
+                EdmTypeKind.Primitive when edmTypeReference.IsBinary() => Convert.ToBase64String(new byte[] { 0x00 }),
+                EdmTypeKind.Primitive when edmTypeReference.IsBoolean() => true,
+                EdmTypeKind.Primitive when edmTypeReference.IsByte() => 0x00,
+                EdmTypeKind.Primitive when edmTypeReference.IsDate() => DateTime.MinValue.ToString("o", CultureInfo.InvariantCulture),
+                EdmTypeKind.Primitive when edmTypeReference.IsDateTimeOffset() => DateTimeOffset.MinValue.ToString("o", CultureInfo.InvariantCulture),
+                EdmTypeKind.Primitive when edmTypeReference.IsGuid() => Guid.Empty.ToString("D", CultureInfo.InvariantCulture),
+                EdmTypeKind.Primitive when edmTypeReference.IsInt16() ||
+                                            edmTypeReference.IsInt32() ||
+                                            edmTypeReference.IsDecimal() ||
+                                            edmTypeReference.IsInt64() ||
+                                            edmTypeReference.IsFloating() ||
+                                            edmTypeReference.IsDouble() => 0,
+                EdmTypeKind.Primitive => GetTypeNameForPrimitive(context, edmTypeReference, document),
+
+                EdmTypeKind.Entity or EdmTypeKind.Complex or EdmTypeKind.Enum => new JsonObject()
+                    {//TODO this is wrong for enums, and should instead use one of the enum members
```

Comment:
Just double checking if this TODO is relevant? The functionality seems the same after the refactor

Reply:
good catch I had completely forgotten about this!

No, this is incorrect, the example for an enum ends up being

```json
{
   "@odata.type": "foo"
}
```

which does not validate with tools like spectral and should be instead

```json
"FooMember1"
```

I'll push a patch soon

Reply:
patched